docs/developer/architecture/interaction-with-k8s-from-salt.rst

```diff
@@ -0,0 +1,96 @@
+Interaction with Kubernetes API from Salt
+=========================================
+
+Context
+-------
+
+In MetalK8s we need to interact with Kubernetes API from Salt in order to
+create and edit various objects. Since the Salt Kubernetes module is not
+flexible enough (at least today) we decided to build our own Salt module that
+will cover all we need in MetalK8s. This Salt module relies on python
+Kubernetes lib so that we have a proper python class for every Kubernetes
+object and we can validate the object content without any call to the actual
+Kubernetes API.
+
+But Python Kubernetes lib is a bit late on Kubernetes versions (at the
+time of writing, the latest python-kubernetes is for k8s-1.18 and k8s 1.22.1
+just came out) so it means that some Kubernetes objects from the latest
+versions cannot be managed by this python lib.
+
+Here, we want to solve this issue and make sure when a new Kubernetes
+version comes out that we can easily support interaction with all the
+Kubernetes API from Salt.
+
+Design Choices
+--------------
+
+To have proper support of the full Kubernetes API we want to use directly
+``kubectl``, which is the CLI build as part of Kubernetes.
+
+The advantage of using ``kubectl`` is that we do not have anything to do in
+order to support interaction with the new Kubernetes API and nothing (or
+almost) to maintain in our Salt module that uses this CLI.
+It also means that we need to use a CLI in order to do some “HTTP query”
+to Kubernetes API, which adds an extra layer, but it’s a small inconvenience
+compare to the benefits.
+
+Rejected Design Choices
+-----------------------
+
+Contribute to python-kubernetes Lib
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We could contribute to python-kubernetes in order to make it compatible with
+the latest versions of Kubernetes.
+
+This approach was rejected because python-kubernetes start to be really late
+on Kubernetes versions, and this work, likely, needs to be done for every new
+Kubernetes version.
```

Comment:
This would still be my preference. Yes, it adds a bit of work, but that's work we could/should take on/help out with upstream, for everyone's benefit. Resorting to invoking `kubectl` from a Python app because the upstream Python client bindings struggle to keep up with new releases of the server doesn't feel right to me.